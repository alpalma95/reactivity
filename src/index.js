import { h } from "./factories/createElement.js";
import { stream, hook } from "./streams.js";
import { $ } from "./factories/select.js";
import { safeRemove } from "./factories/effectsManager.js";
import { registerDirective } from "./directives/registerDirectives.js";

export { h, stream, hook, $, safeRemove, registerDirective };
