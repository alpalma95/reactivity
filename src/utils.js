import { hook } from "./streams.js";

// TODO: Global WeakMap with all needed effects. We shall look for a better place
/** @type {WeakMap<HTMLElement, Array>} */
const DOMEffectsMap = new WeakMap();

export const registerEffect = (element, effect) =>
    DOMEffectsMap.has(element)
        ? DOMEffectsMap.get(element).push(effect)
        : DOMEffectsMap.set(element, [effect]);

/**
 * Remove the given element as well as the effects associated with it found in the weakmap
 * @param {HTMLElement} Element Element to remove from the DOM
 */
export const safeRemove = (element) => {
    if (typeof element.destroy === "function") element.destroy(element);

    DOMEffectsMap.get(element)?.forEach((effect) => {
        Array.isArray(effect)
            ? effect.forEach((nestedEffect) => nestedEffect.unhook())
            : effect.unhook();
    });
    DOMEffectsMap.delete(element);
    element?.remove();
};

export const bindAttribute = (element, attrName, cb) => {
    let effect = hook(() => element.setAttribute(attrName, cb(element)));
    registerEffect(element, effect);
};

export const switchProps = (props, children) => {
    if (Array.isArray(props)) {
        children = props;
        props = {};
    }
    return [props, children];
};

/**
 *
 * @param {HTMLElement} root
 * @returns {Array<HTMLElement>}
 */
export const getRefs = (root, selector) => {
    const refs = [];
    /** @type {TreeWalker} */
    const walker = document.createTreeWalker(
        root,
        NodeFilter.SHOW_ELEMENT,
        function (node) {
            if (node.getAttribute("ref") == selector)
                return NodeFilter.FILTER_ACCEPT;
            if (node.getAttribute("ref")?.toUpperCase().includes("CONTROLLER"))
                return NodeFilter.FILTER_REJECT;

            return NodeFilter.FILTER_SKIP;
        }
    );

    let currentNode;
    while ((currentNode = walker.nextNode())) refs.push(currentNode);

    return refs;
};
