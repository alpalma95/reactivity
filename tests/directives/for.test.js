import { h, stream } from "../../src";

describe("for directive", () => {
    const items = stream([
        { id: 1, name: "one" },
        { id: 2, name: "two" },
        { id: 3, name: "three" },
    ]);
    const template = (item) => h.li({ "data-key": item.id }, [item.name]);
    const element = h.ul({
        "rv-for": [items, template],
        "data-track-by": "id",
    });

    beforeEach(() => {
        items.val = [
            { id: 1, name: "one" },
            { id: 2, name: "two" },
            { id: 3, name: "three" },
        ];
    });

    it("should render the items", () => {
        expect(element.children.length).toBe(items.val.length);
        [...element.children].forEach((child, index) => {
            expect(+child.dataset.key).toBe(items.val[index].id);
            expect(child.innerHTML).toBe(items.val[index].name);
        });
    });

    it("should append a new child at the end", () => {
        items.val = [...items.val, { id: 4, name: "four" }];
        expect(element.children.length).toBe(items.val.length);
        expect(element.children[3].innerHTML).toBe("four");
    });

    it("should delete a child without altering the order", () => {
        items.val = [...items.val.filter((item) => item.id !== 2)];
        expect(element.children.length).toBe(items.val.length);
        expect(element.querySelector('[data-key="2"]')).toBeNull();

        [...element.children].forEach((child, index) => {
            expect(+child.dataset.key).toBe(items.val[index].id);
            expect(child.innerHTML).toBe(items.val[index].name);
        });
    });

    it("should sync with the order of the array if it changes completely", () => {
        items.val = [
            ...items.val.sort((person1, person2) => person2.id - person1.id),
        ];
        [...element.children].forEach((child, index) => {
            expect(+child.dataset.key).toBe(items.val[index].id);
            expect(child.innerHTML).toBe(items.val[index].name);
        });
    });

    it("should insert the new child at the right position", () => {
        const left = items.val.filter((item) => item.id < 3);
        const right = items.val.filter((item) => item.id >= 3);

        items.val = [
            ...left,
            { id: 89, name: "inserted in the middle" },
            ...right,
        ];

        [...element.children].forEach((child, index) => {
            expect(+child.dataset.key).toBe(items.val[index].id);
            expect(child.innerHTML).toBe(items.val[index].name);
        });
    });

    it("should work with nested loops", () => {
        const items = stream([
            { id: 1, name: "one" },
            {
                id: 2,
                name: "two",
                children: stream([
                    { id: 4, name: "four" },
                    { id: 5, name: "five" },
                    { id: 6, name: "six" },
                ]),
            },
            { id: 3, name: "three" },
        ]);

        const template = (item) => {
            return h.li({ "data-key": item.id }, [
                item.name,
                item.children
                    ? h.ul({
                          "rv-for": [item.children, template],
                          "data-track-by": "id",
                      })
                    : undefined,
            ]);
        };

        const Test = () => {
            return h.ul({ "rv-for": [items, template], "data-track-by": "id" });
        };

        expect(Test()).toMatchSnapshot("./snapshots/for/nested.output.html");

        items.val.find((item) => item.id === 2).children.val = [
            { id: 7, name: "seven" },
        ];

        expect(Test()).toMatchSnapshot(
            "./snapshots/for/nestedModified.output.html"
        );
    });

    it("should transform the data if transformFn is provided", () => {
        const nums = stream([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
        const template = (item) => h.li({ "data-key": item.val }, [item.val]);
        const Test = () => {
            return h.ul({
                "data-track-by": "val",
                "rv-for": [nums, template, (n) => ({ val: n })],
            });
        };
        expect(Test()).toMatchSnapshot(
            "./snapshots/for/transformFn.output.html"
        );
    });
});
