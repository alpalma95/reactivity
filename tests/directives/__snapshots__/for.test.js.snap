// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`for directive > should keep on transforming the data if transformFn is provided > ./snapshots/for/transformFnUpdate.output.html 1`] = `
<ul
  data-track-by="val"
>
  <li
    data-key="1"
  >
    1
     at index 0
  </li>
  <li
    data-key="2"
  >
    2
     at index 1
  </li>
  <li
    data-key="3"
  >
    3
     at index 2
  </li>
  <li
    data-key="4"
  >
    4
     at index 3
  </li>
  <li
    data-key="5"
  >
    5
     at index 4
  </li>
  <li
    data-key="6"
  >
    6
     at index 5
  </li>
  <li
    data-key="7"
  >
    7
     at index 6
  </li>
  <li
    data-key="8"
  >
    8
     at index 7
  </li>
  <li
    data-key="9"
  >
    9
     at index 8
  </li>
  <li
    data-key="10"
  >
    10
     at index 9
  </li>
  <li
    data-key="11"
  >
    11
     at index 10
  </li>
</ul>
`;

exports[`for directive > should transform the data if transformFn is provided > ./snapshots/for/transformFn.output.html 1`] = `
<ul
  data-track-by="val"
>
  <li
    data-key="1"
  >
    1
  </li>
  <li
    data-key="2"
  >
    2
  </li>
  <li
    data-key="3"
  >
    3
  </li>
  <li
    data-key="4"
  >
    4
  </li>
  <li
    data-key="5"
  >
    5
  </li>
  <li
    data-key="6"
  >
    6
  </li>
  <li
    data-key="7"
  >
    7
  </li>
  <li
    data-key="8"
  >
    8
  </li>
  <li
    data-key="9"
  >
    9
  </li>
  <li
    data-key="10"
  >
    10
  </li>
</ul>
`;

exports[`for directive > should work with nested loops > ./snapshots/for/nested.output.html 1`] = `
<ul
  data-track-by="id"
>
  <li
    data-key="1"
  >
    one
    
  </li>
  <li
    data-key="2"
  >
    two
    <ul
      data-track-by="id"
    >
      <li
        data-key="4"
      >
        four
        
      </li>
      <li
        data-key="5"
      >
        five
        
      </li>
      <li
        data-key="6"
      >
        six
        
      </li>
    </ul>
  </li>
  <li
    data-key="3"
  >
    three
    
  </li>
</ul>
`;

exports[`for directive > should work with nested loops > ./snapshots/for/nestedModified.output.html 1`] = `
<ul
  data-track-by="id"
>
  <li
    data-key="1"
  >
    one
    
  </li>
  <li
    data-key="2"
  >
    two
    <ul
      data-track-by="id"
    >
      <li
        data-key="7"
      >
        seven
        
      </li>
    </ul>
  </li>
  <li
    data-key="3"
  >
    three
    
  </li>
</ul>
`;
